http/backup_service.go

```diff
@@ -48,131 +40,66 @@ type BackupHandler struct {
+func (h *BackupHandler) handleBackupKVStore(w http.ResponseWriter, r *http.Request) {
+	span, r := tracing.ExtractFromHTTPRequest(r, "BackupHandler.handleBackupKVStore")
 	defer span.Finish()
 
 	ctx := r.Context()
 
-	id, files, err := h.BackupService.CreateBackup(ctx)
-	if err != nil {
-		h.HandleHTTPError(ctx, err, w)
-		return
-	}
-
-	internalBackupPath := h.BackupService.InternalBackupPath(id)
-
-	boltPath := filepath.Join(internalBackupPath, bolt.DefaultFilename)
-	boltFile, err := os.OpenFile(boltPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0660)
-	if err != nil {
-		err = multierr.Append(err, os.RemoveAll(internalBackupPath))
-		h.HandleHTTPError(ctx, err, w)
-		return
-	}
-
-	if err = h.KVBackupService.Backup(ctx, boltFile); err != nil {
-		err = multierr.Append(err, os.RemoveAll(internalBackupPath))
-		h.HandleHTTPError(ctx, err, w)
-		return
-	}
-
-	files = append(files, bolt.DefaultFilename)
-
-	credsExist, err := h.backupCredentials(internalBackupPath)
-
-	if err != nil {
+	if err := h.BackupService.BackupKVStore(ctx, w); err != nil {
 		h.HandleHTTPError(ctx, err, w)
 		return
 	}
-
-	if credsExist {
-		files = append(files, fs.DefaultConfigsFile)
-	}
-
-	b := backup{
-		ID:    id,
-		Files: files,
-	}
-	if err = json.NewEncoder(w).Encode(&b); err != nil {
-		err = multierr.Append(err, os.RemoveAll(internalBackupPath))
-		h.HandleHTTPError(ctx, err, w)
-		return
-	}
-}
-
-func (h *BackupHandler) backupCredentials(internalBackupPath string) (bool, error) {
-	credBackupPath := filepath.Join(internalBackupPath, fs.DefaultConfigsFile)
-
-	credPath, err := defaultConfigsPath()
-	if err != nil {
-		return false, err
-	}
-	token, err := ioutil.ReadFile(credPath)
-	if err != nil && !os.IsNotExist(err) {
-		return false, err
-	} else if os.IsNotExist(err) {
-		return false, nil
-	}
-
-	if err := ioutil.WriteFile(credBackupPath, []byte(token), 0600); err != nil {
-		return false, err
-	}
-	return true, nil
 }
 
-func (h *BackupHandler) handleFetchFile(w http.ResponseWriter, r *http.Request) {
-	span, r := tracing.ExtractFromHTTPRequest(r, "BackupHandler.handleFetchFile")
+func (h *BackupHandler) handleBackupShard(w http.ResponseWriter, r *http.Request) {
+	span, r := tracing.ExtractFromHTTPRequest(r, "BackupHandler.handleBackupShard")
 	defer span.Finish()
 
 	ctx := r.Context()
 
 	params := httprouter.ParamsFromContext(ctx)
-	backupID, err := strconv.Atoi(params.ByName("backup_id"))
+	shardID, err := strconv.ParseUint(params.ByName("shardID"), 10, 64)
 	if err != nil {
 		h.HandleHTTPError(ctx, err, w)
 		return
 	}
-	backupFile := params.ByName("backup_file")
 
-	if err = h.BackupService.FetchBackupFile(ctx, backupID, backupFile, w); err != nil {
+	var since time.Time
+	if s := r.URL.Query().Get("since"); s != "" {
+		if since, err = time.ParseInLocation(time.RFC3339, s, time.UTC); err != nil {
+			h.HandleHTTPError(ctx, err, w)
+			return
+		}
+	}
+
+	if err := h.BackupService.BackupShard(ctx, w, shardID, since); err != nil {
```

Comment:
Would it be helpful here if `err` was of type `*influxdb.Error` and the `Code` was `ENotFound`:

https://github.com/influxdata/influxdb/blob/9c4dfed391fba2f5e433ebd5d14176942dd53e81/errors.go#L18

That would allow the HTTP handler to return a `404` for this situation. If the backup client expected 404s, the backup CLI tool could safely skip the error and continue the backup. This would account for the racy situation where a shard was deleted by the retention policy service or some other action, whilst a backup was occurring.